### ISSUE TITLE
Trident can get stuck in CrashLoopBackoff if it is restarted while processing a volume delete

### ISSUE THREAD

**Issue by dutchiechris** [NetApp/trident#32]
There seems to be a timing issue whereby if Trident restarts (upgrade or otherwise) while it was processing a PV delete it can get stuck in a `CrashLoopBackoff`.

I reproduced with a SolidFire backend by:
1. Creating 300 PVCs which causes Trident to create each of one
1. Delete 300 PVCs which causes Trident to delete each one
    1. While deletes are running (each will take a second or two) delete the trident pod (ex: `kubectl delete po trident-3831985011-3h5v6`)
1. It might take a couple of `delete` tries to encounter the failure, but eventually you will see the pod get stuck in a `CrashLoopBackoff`

At that point logs show something like this on each restart retry:
```
time="2017-06-20T19:41:36Z" level=info msg="Processed volume transaction log." op=deleteVolume size=1073741824 storageClass=bronze volume=loadgen-pvc-loadgen-114-6e248
time="2017-06-20T19:41:36Z" level=info msg="Volume for delete transaction found." name=loadgen-pvc-loadgen-114-6e248
time="2017-06-20T19:41:36Z" level=error msg="unable to locate volume for delete operation: volume not found"
time="2017-06-20T19:41:36Z" level=error msg="Unable to delete volume from backend." backend="solidfire_192.168.1.190" volume=loadgen-pvc-loadgen-114-6e248
time="2017-06-20T19:41:36Z" level=fatal msg="Could not bootstrap from persistent store! Bootstrap might have failed, persistent store might be down, or persistent store may not have any backend, volume, or storage class state: Unable to clean up deleted volume loadgen-pvc-loadgen-114-6e248:  volume not found"
```

I checked on the SolidFire and saw the device it was complaining about was in the `Deleted` volume list (the only volume in that list!) so I chose to `Restore` it.  After that Trident was able to restart successfully and proceeded to delete all the PVs that were no longer needed as it should.  It seems odd that the `Deleted` volume list only had a single volume, the one causing the issue.  Perhaps that will help locate the bug.

Also, if this kind of situation is encountered (one PV of potentially thousands) it would be better as an 'error' rather than a 'fatal' level message that stops all provisioning.

**Comment by kangarlou** [comment]
Thanks Chris! The problem is that the underlying nDVP drivers don't have a consistent behavior during volume deletion. ontap-nas and ontap-san drivers return no error when prompted to delete a nonexistent volume whereas the SolidFire driver returns an error: https://github.com/NetApp/netappdvp/blob/master/storage_drivers/solidfire_san.go#L290

Your observation that the deleted volume list had only a single volume makes sense as Trident serializes all operations. I'm a little hesitant to let bootstrapping proceed during such failures as this sort of recovery requires manual intervention and chances are admins may not take corrective actions **_even if_** such problems are highlighted during bootstrapping. Having said that, I think Trident should do a better job of reporting why a failure happened and what corrective action needs to be taken in order to recover.

**Comment by kangarlou** [comment]
Chris, could you please run your test with an ontap-nas or ontap-san backend and report the results? I think the problem is limited to the SolidFire driver and the fix doesn't require any changes to Trident. Thanks!

**Comment by dutchiechris** [comment]
@kangarlou 

I just ran with an ONTAP NAS backend and saw this:
```
time="2017-06-21T14:25:29Z" level=info msg="Processed volume transaction log." op=addVolume size=1073741824 storageClass=silver-shared volume=loadgen-pvc-loadgen-10-818fa
time="2017-06-21T14:25:29Z" level=warning msg="Volume already deleted while destroying volume: trident_loadgen_pvc_loadgen_10_818fa
resultStatusAttr: failed
resultReasonAttr: Volume "trident_loadgen_pvc_loadgen_10_818fa" does not exist in Vserver "cmdemo_docker". Reason: entry doesn't exist.
resultErrnoAttr: 13040
error: <nil>"
time="2017-06-21T14:25:29Z" level=error msg="unable to locate volume for delete operation: volume not found"
time="2017-06-21T14:25:29Z" level=fatal msg="Could not bootstrap from persistent store! Bootstrap might have failed, persistent store might be down, or persistent store may not have any backend, volume, or storage class state: Error attempting to clean up volume loadgen-pvc-loadgen-10-818fa from backend solidfire_192.168.1.190:  volume not found"
```

What is quite odd is the warning msg quite clearly comes from ONTAP yet the final error claims it isn't in the solidfire backend?  Odd.  I just created that vol on the ONTAP system and deleted the trident pod and it restarted and logged the same errors as above again, meanwhile it deleted that vol on ONTAP.  I'm not sure how I can get trident to start and stay up to do more testing.  Ideas?

**Comment by kangarlou** [comment]
The reason you're hitting the same problem as before is because you haven't cleared the state associated with the problematic volume in etcd. Here is how to do it:

1. Bring up the etcd container by itself:
```
$ cat etcd-pod.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: etcd
  labels:
    app: etcd
spec:
  containers:
  - name: etcd
    image: quay.io/coreos/etcd:v3.1.3
    command:
    - /usr/local/bin/etcd
    args:
     - -name
     - etcd1
     - -advertise-client-urls
     - http://127.0.0.1:8001
     - -listen-client-urls
     - http://127.0.0.1:8001
     - -initial-cluster
     - default=http://127.0.0.1:8002
     - -initial-advertise-peer-urls
     - http://127.0.0.1:8002
     - -listen-peer-urls
     - http://127.0.0.1:8002
     - -data-dir
     - /var/etcd/data
     - -initial-cluster
     - etcd1=http://127.0.0.1:8002
    volumeMounts:
    - name: etcd-vol
      mountPath: /var/etcd/data
  volumes:
      - name: etcd-vol
        persistentVolumeClaim:
          claimName: trident
$ kubectl create -f etcd-pod.yaml
```
2. Find the outstanding transaction at the time of the failure:

`$ kubectl exec etcd --  /usr/local/bin/etcdctl --endpoints http://127.0.0.1:8001 ls trident/v1/txn`

In your case, you should see `loadgen-pvc-loadgen-114-6e248` returned.

3. Delete the state corresponding to the volume transaction:

`$ kubectl exec etcd --  /usr/local/bin/etcdctl --endpoints http://127.0.0.1:8001 rm trident/v1/txn/loadgen-pvc-loadgen-114-6e248`

4. Bring down the etcd pod, restart the Trident deployment configured with an ONTAP backend, and rerun your tests.

Trident's bootstrapping code takes care of the outstanding volume transaction; however, due to the problem with the SolidFire driver, it couldn't get past this stage.

Please let me know if you encounter any problems with an ONTAP backend. If everything looks fine, I'll push my fix to the SolidFire driver to the nDVP repo. Thanks!

**Comment by dutchiechris** [comment]
Using the recipe above I inserted a step 0. `kubectl scale deployment trident --replicas=0` to make sure trident and etcd would stay down to do the etcd edit safely, and then added step 5. with the same and `--replicas=1` to bring it back up.  Also, for clarity, in step 4. do `kubectl delete -f etcd-pod.yaml` to bring the etcd down.  This recipe worked fine and trident started successfullly.

Researching my earlier logs with an ONTAP txn entry and a SolidFire backend error message we discovered it was due to code [here](https://github.com/NetApp/trident/blob/v17.07.0-beta.1/core/orchestrator_core.go#L241).  Basically if a provisioning/deprovisioning makes it part-way through the backend is not known so all are tried, and since I have a SolidFire one defined we could hit that bug in the ONTAP case too.

Thanks @kangarlou for troubleshooting and fixing so quickly!  Next is to make sure it gets included in 17.07.0 when it ships...
